File: t/VB/OmrProcessingDemo/Operations/DetectBarcodeOperation.vb
Imports Leadtools
Imports Leadtools.Barcode
Imports Leadtools.Forms.Common

Friend Class DetectBarcodeOperation
   Inherits BusyOperation

   Public Property DetectedSymbology As BarcodeSymbology
   Public Property IsSymbologyDetected As Boolean
   Private image As RasterImage
   Private rect As LeadRect

   Public Sub New(ByVal loadedImage As RasterImage, ByVal lr As LeadRect)
      MyBase.New()
      image = loadedImage
      rect = lr
   End Sub

   Protected Overrides Sub Run()
      Progress(33, "Creating components")
      Dim engine As BarcodeEngine = New BarcodeEngine()
      Dim options As BarcodeReadOptions() = GetHorizontalAndVerticalReadBarcodeOptions(engine.Reader)
      Progress(66, "Detecting barcode symbology")
      Dim bc As BarcodeData() = engine.Reader.ReadBarcodes(image, rect, 1, engine.Reader.GetAvailableSymbologies(), options)

      If bc IsNot Nothing AndAlso bc.Length > 0 Then
         DetectedSymbology = bc(0).Symbology
         IsSymbologyDetected = True
      Else
         IsSymbologyDetected = False
      End If

      MyBase.Run()
   End Sub

   Private Shared Function GetHorizontalAndVerticalReadBarcodeOptions(ByVal reader As BarcodeReader) As BarcodeReadOptions()
      Dim oneDReadOptions As OneDBarcodeReadOptions = TryCast(reader.GetDefaultOptions(BarcodeSymbology.UPCA).Clone(), OneDBarcodeReadOptions)
      oneDReadOptions.SearchDirection = BarcodeSearchDirection.HorizontalAndVertical
      Dim fourStateReadOptions As FourStateBarcodeReadOptions = TryCast(reader.GetDefaultOptions(BarcodeSymbology.USPS4State).Clone(), FourStateBarcodeReadOptions)
      fourStateReadOptions.SearchDirection = BarcodeSearchDirection.HorizontalAndVertical
      Dim postNetPlanetReadOptions As PostNetPlanetBarcodeReadOptions = TryCast(reader.GetDefaultOptions(BarcodeSymbology.PostNet).Clone(), PostNetPlanetBarcodeReadOptions)
      postNetPlanetReadOptions.SearchDirection = BarcodeSearchDirection.HorizontalAndVertical
      Dim gs1StackedReadOptions As GS1DatabarStackedBarcodeReadOptions = TryCast(reader.GetDefaultOptions(BarcodeSymbology.GS1DatabarStacked).Clone(), GS1DatabarStackedBarcodeReadOptions)
      gs1StackedReadOptions.SearchDirection = BarcodeSearchDirection.HorizontalAndVertical
      Dim patchCodeReadOptions As PatchCodeBarcodeReadOptions = TryCast(reader.GetDefaultOptions(BarcodeSymbology.PatchCode).Clone(), PatchCodeBarcodeReadOptions)
      patchCodeReadOptions.SearchDirection = BarcodeSearchDirection.HorizontalAndVertical
      Dim pdf417ReadOptions As PDF417BarcodeReadOptions = TryCast(reader.GetDefaultOptions(BarcodeSymbology.PDF417).Clone(), PDF417BarcodeReadOptions)
      pdf417ReadOptions.SearchDirection = BarcodeSearchDirection.HorizontalAndVertical
      Dim microPdf417ReadOptions As MicroPDF417BarcodeReadOptions = TryCast(reader.GetDefaultOptions(BarcodeSymbology.MicroPDF417).Clone(), MicroPDF417BarcodeReadOptions)
      microPdf417ReadOptions.SearchDirection = BarcodeSearchDirection.HorizontalAndVertical
      Dim readOptions As BarcodeReadOptions() = {oneDReadOptions, fourStateReadOptions, postNetPlanetReadOptions, gs1StackedReadOptions, patchCodeReadOptions, pdf417ReadOptions, microPdf417ReadOptions}
      Return readOptions
   End Function
End Class

File: t/CS/OmrProcessingDemo/Operations/DetectBarcodeOperation.cs

using Leadtools;
using Leadtools.Barcode;
using Leadtools.Forms.Common;

namespace OmrProcessingDemo.Operations
{
   internal class DetectBarcodeOperation : BusyOperation
   {
      public BarcodeSymbology DetectedSymbology { get; private set; }
      public bool IsSymbologyDetected { get; private set; }

      RasterImage image;
      LeadRect rect;
      public DetectBarcodeOperation(RasterImage loadedImage, LeadRect lr) : base()
      {
         image = loadedImage;
         rect = lr;
      }

      protected override void Run()
      {
         Progress(33, "Creating components");
         BarcodeEngine engine = new BarcodeEngine();
         BarcodeReadOptions[] options = GetHorizontalAndVerticalReadBarcodeOptions(engine.Reader);

         Progress(66, "Detecting barcode symbology");
         BarcodeData[] bc = engine.Reader.ReadBarcodes(image, rect, 1, engine.Reader.GetAvailableSymbologies(), options);

         if (bc != null && bc.Length > 0)
         {
            DetectedSymbology = bc[0].Symbology;
            IsSymbologyDetected = true;
         }
         else
         {
            IsSymbologyDetected = false;
         }

         base.Run();
      }

      private static BarcodeReadOptions[] GetHorizontalAndVerticalReadBarcodeOptions(BarcodeReader reader)
      {
         // By default, the options read horizontal barcodes only, create an array of options capable of reading vertical barcodes 

         // Notice, we cloned the default options in reader so we will not change the original options 

         OneDBarcodeReadOptions oneDReadOptions = reader.GetDefaultOptions(BarcodeSymbology.UPCA).Clone() as OneDBarcodeReadOptions;
         oneDReadOptions.SearchDirection = BarcodeSearchDirection.HorizontalAndVertical;

         FourStateBarcodeReadOptions fourStateReadOptions = reader.GetDefaultOptions(BarcodeSymbology.USPS4State).Clone() as FourStateBarcodeReadOptions;
         fourStateReadOptions.SearchDirection = BarcodeSearchDirection.HorizontalAndVertical;

         PostNetPlanetBarcodeReadOptions postNetPlanetReadOptions = reader.GetDefaultOptions(BarcodeSymbology.PostNet).Clone() as PostNetPlanetBarcodeReadOptions;
         postNetPlanetReadOptions.SearchDirection = BarcodeSearchDirection.HorizontalAndVertical;

         GS1DatabarStackedBarcodeReadOptions gs1StackedReadOptions = reader.GetDefaultOptions(BarcodeSymbology.GS1DatabarStacked).Clone() as GS1DatabarStackedBarcodeReadOptions;
         gs1StackedReadOptions.SearchDirection = BarcodeSearchDirection.HorizontalAndVertical;

         PatchCodeBarcodeReadOptions patchCodeReadOptions = reader.GetDefaultOptions(BarcodeSymbology.PatchCode).Clone() as PatchCodeBarcodeReadOptions;
         patchCodeReadOptions.SearchDirection = BarcodeSearchDirection.HorizontalAndVertical;

         PDF417BarcodeReadOptions pdf417ReadOptions = reader.GetDefaultOptions(BarcodeSymbology.PDF417).Clone() as PDF417BarcodeReadOptions;
         pdf417ReadOptions.SearchDirection = BarcodeSearchDirection.HorizontalAndVertical;

         MicroPDF417BarcodeReadOptions microPdf417ReadOptions = reader.GetDefaultOptions(BarcodeSymbology.MicroPDF417).Clone() as MicroPDF417BarcodeReadOptions;
         microPdf417ReadOptions.SearchDirection = BarcodeSearchDirection.HorizontalAndVertical;

         // Even though this array will not contain all options, it should be enough to read all barcodes, since the version of ReadBarcodes we will use 
         // will use the default options if an override is not passed 
         BarcodeReadOptions[] readOptions =
         {
            oneDReadOptions, fourStateReadOptions, postNetPlanetReadOptions, gs1StackedReadOptions, patchCodeReadOptions, pdf417ReadOptions, microPdf417ReadOptions
         };

         return readOptions;
      }
   }
}